internal/charmstore/resources.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charmstore // import "gopkg.in/juju/charmstore.v5-unstable/internal/charmstore"
+import (
+	"fmt"
+	"io"
+	"sort"
+	"time"
+
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+)
+
+// newResourceQuery returns a mongo query doc that will retrieve the
+// given named resource and revision associated with the given charm URL.
+// If revision is < 0, all revisions of the resource will be selected by
+// the query.
+func newResourceQuery(url *charm.URL, name string, revision int) bson.D {
+	query := make(bson.D, 2, 3)
+	query[0] = bson.DocElem{"baseurl", mongodoc.BaseURL(url)}
+	query[1] = bson.DocElem{"name", name}
+	if revision >= 0 {
+		query = append(query, bson.DocElem{"revision", revision})
+	}
+	return query
+}
+
+// sortResources sorts the provided resource docs, The resources are
+// sorted first by URL then by name and finally by revision.
+func sortResources(resources []*mongodoc.Resource) {
+	sort.Sort(resourcesByName(resources))
+}
+
+type resourcesByName []*mongodoc.Resource
+
+func (sorted resourcesByName) Len() int      { return len(sorted) }
+func (sorted resourcesByName) Swap(i, j int) { sorted[i], sorted[j] = sorted[j], sorted[i] }
+func (sorted resourcesByName) Less(i, j int) bool {
+	r0, r1 := sorted[i], sorted[j]
+	if *r0.BaseURL != *r1.BaseURL {
+		return r0.BaseURL.String() < r1.BaseURL.String()
+	}
+	if r0.Name != r1.Name {
+		return r0.Name < r1.Name
+	}
+	return r0.Revision < r1.Revision
+}
+
+var resourceNotFound = errgo.Newf("resource not found")
+
+// ListResources returns the set of resources for the charm. If the
+// unpublished channel is specified then set is composed of the latest
+// revision for each resource. Otherwise it holds the revisions declared
+// when the charm/channel pair was published.
+func (s *Store) ListResources(entity *mongodoc.Entity, channel params.Channel) ([]*mongodoc.Resource, error) {
+	if entity.URL.Series == "bundle" {
+		return nil, errgo.Newf("bundles do not have resources")
+	}
+	if entity.CharmMeta == nil {
+		return nil, errgo.Newf("entity missing charm metadata")
+	}
+	baseEntity, err := s.FindBaseEntity(entity.URL, FieldSelector("channelresources"))
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	var docs []*mongodoc.Resource
+	for name, meta := range entity.CharmMeta.Resources {
+		doc, err := s.latestResource(baseEntity, channel, name)
```

Comment:
When the channel is unpublished, presumably this *could* be more efficient than making a query for each resource - it could do a single query on the resources collection for all resources with a matching URL and filter out all those that aren't mentioned in the charm metadata. If you don't fancy doing that now, perhaps add a TODO?